packet_encoder.go

```diff
@@ -22,6 +22,13 @@ type packetEncoder interface {
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// Statistics in order to expose metrics
+	offset() int
+	recordTopic(topic string)
```

Comment:
this isn't at the right level; the protocol-level encoding/decoding shouldn't have any knowledge of semantic values like topic names